config/src/main/java/org/springframework/security/config/annotation/web/configurers/X509Configurer.java

```diff
@@ -177,8 +179,16 @@ public X509Configurer<H> subjectPrincipalRegex(String subjectPrincipalRegex) {
 	public void init(H http) {
 		PreAuthenticatedAuthenticationProvider authenticationProvider = new PreAuthenticatedAuthenticationProvider();
 		authenticationProvider.setPreAuthenticatedUserDetailsService(getAuthenticationUserDetailsService(http));
+		authenticationProvider.setGrantedAuthoritySupplier(() -> AuthorityUtils.createAuthorityList("FACTOR_X509"));
 		http.authenticationProvider(authenticationProvider)
 			.setSharedObject(AuthenticationEntryPoint.class, new Http403ForbiddenEntryPoint());
+		ExceptionHandlingConfigurer<H> exceptions = http.getConfigurer(ExceptionHandlingConfigurer.class);
+		if (exceptions != null) {
+			AuthenticationEntryPoint forbidden = new Http403ForbiddenEntryPoint();
+			exceptions.defaultAuthenticationEntryPointFor(forbidden, AnyRequestMatcher.INSTANCE);
```

Comment:
I'm not sure about this change.

If you have X509 invoked first then the default `AuthenticationEntryPoint` is forbidden for everything. What happens if X509 and password are required and a user is authenticated with X509 but not password? I'd expect that the result would be prompt for a password but it appears that this will result in forbidden.

Reply:
If the user is authenticated with X509, but not password, then this `AuthenticationEntryPoint` is not exercised. Instead the one configured in the missing-authorities handler is.

I think this speaks to your other comments about the method names being confusing.

Reply:
Either way, I think [it might fix a bug](https://github.com/spring-projects/spring-security/blob/main/config/src/main/java/org/springframework/security/config/annotation/web/configurers/X509Configurer.java#L184). It doesn't have to do with the PR, and I will log it separately.